Work around metal buffer bug on MacOS + AMD GPU

Kethen opened the pull request:
~With VMA_MEMORY_USAGE_CPU_TO_GPU buffers, metal buffer appears 0 filled in metal trace during vkCmdCopyBufferToImage triggered MTLBlitCommandEncoder instance method.~

~Allocate VMA_MEMORY_USAGE_GPU_TO_CPU instead on MacOS + AMD GPU~

Update: adding VK_MEMORY_PROPERTY_HOST_COHERENT_BIT to allocation on MacOS dGPU, ref https://github.com/KhronosGroup/MoltenVK/issues/960

This PR reenables vulkan on intel Macs, probably make sense to gather everyone from these issues to test it before considering merging:

https://github.com/hrydgard/ppsspp/issues/15473
https://github.com/hrydgard/ppsspp/issues/16990
https://github.com/hrydgard/ppsspp/issues/18139
https://github.com/hrydgard/ppsspp/issues/19594
...

Kethen committed:
Work around metal buffer bug on MacOS + AMD GPU

With VMA_MEMORY_USAGE_CPU_TO_GPU buffers, metal buffer appears 0
filled in metal trace during vkCmdCopyBufferToImage triggered
MTLBlitCommandEncoder instance method.

Allocate VMA_MEMORY_USAGE_GPU_TO_CPU instead on MacOS + AMD GPU

Kethen commented:
Hi @eks28 @XStylus @setanarut @Kifary @Halo-Michael @jerrypc2036

Please test the build artifact of https://github.com/hrydgard/ppsspp/actions/runs/12225492690 if possible

Halo-Michael commented:
> Hi @eks28 @XStylus @setanarut @Kifary @Halo-Michael @jerrypc2036 @fat4talk
> 
> Please test the build artifact of https://github.com/hrydgard/ppsspp/actions/runs/12225492690 if possible

![ScreenShot2024-12-09 17 33 26](https://github.com/user-attachments/assets/ae0c5012-c9d0-4bc0-9dc4-77e16028be67)
Works fine! Good job!

XStylus commented:
> Please test the build artifact of https://github.com/hrydgard/ppsspp/actions/runs/12225492690 if possible

Did some quick testing of build 25187a9a61 on my system (MacBook Pro 2019 with i9 processor and AMD Radeon Pro 5500M 8 GB). My testing was mainly limited to launching "Final Fantasy Tactics" and carefully scrutinizing the opening video which had lots of motion.

Good news: Vulcan now works in this version, whereas in previous versions (starting with 1.13.0 through 1.18.1) it only presented a black screen, so for starters that's a win.

Less good: OpenGL appears to provide smoother video compared to Vulcan. When watching the FFT opening, there was a very subtle frame skip when in Vulcan mode, whereas there appeared to be none when in OpenGL mode.

Kethen committed:
Switch to VMA_MEMORY_USAGE_CPU_ONLY on MacOS AMD

The ideal allocation would be VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT
like in 1.12.3, but vma picked an index that's actually not mappable

VMA_MEMORY_USAGE_GPU_TO_CPU selects VK_MEMORY_PROPERTY_HOST_CACHED_BIT,
which seems to hurt performance

VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT + VK_MEMORY_PROPERTY_HOST_COHERENT_BIT
selection with VMA_MEMORY_USAGE_CPU_ONLY would suffice for the
workaround, and get better performance

Kethen commented:
Hello ocne again @eks28 @XStylus @setanarut @Kifary @Halo-Michael @jerrypc2036 @fat4talk

I picked another buffer type, please test https://github.com/hrydgard/ppsspp/actions/runs/12234779926

@hrydgard it got more counter intuitive, but asking vma for a
```
VMA_MEMORY_USAGE_GPU_TO_CPU (VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT + VK_MEMORY_PROPERTY_HOST_CACHED_BIT)
```

 gets a slower buffer than
```
VMA_MEMORY_USAGE_CPU_ONLY (VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT + VK_MEMORY_PROPERTY_HOST_COHERENT_BIT)
```

Kethen commented:
adding ref https://github.com/KhronosGroup/MoltenVK/issues/960#issuecomment-663126948

XStylus commented:
Re-tested with revised build in the same manner as I tested the prior build. No perceptible difference detected. Vulcan mode remains functional, albeit still in lesser frame quality compared to OpenGL mode.

Halo-Michael commented:
> Hello ocne again @eks28 @XStylus @setanarut @Kifary @Halo-Michael @jerrypc2036 @fat4talk
> 
> I picked another buffer type, please test https://github.com/hrydgard/ppsspp/actions/runs/12234779926
> 
> @hrydgard it got more counter intuitive, but asking vma for a
> 
> ```
> VMA_MEMORY_USAGE_GPU_TO_CPU (VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT + VK_MEMORY_PROPERTY_HOST_CACHED_BIT)
> ```
> 
> gets a slower buffer than
> 
> ```
> VMA_MEMORY_USAGE_CPU_ONLY (VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT + VK_MEMORY_PROPERTY_HOST_COHERENT_BIT)
> ```

Works too, but I lack large-scale games to test their actual performance.

Kethen committed:
Simplify MacOS AMD GPU workaround

Based on https://github.com/KhronosGroup/MoltenVK/issues/960,
expand MacOS AMD GPU workaround to all dGPU and instead of
changing usage, just append VK_MEMORY_PROPERTY_HOST_COHERENT_BIT

Kethen commented:
> Re-tested with the revised build in the same manner as I tested the prior build. No perceptible difference detected. Vulcan mode remains functional, albeit still in lesser frame quality compared to OpenGL mode.

Thanks for testing! Perhaps https://github.com/hrydgard/ppsspp/actions/runs/12236572156 would help

Kethen commented:
> Re-tested with the revised build in the same manner as I tested the prior build. No perceptible difference detected. Vulcan mode remains functional, albeit still in lesser frame quality compared to OpenGL mode.

@XStylus Thanks for testing! Perhaps https://github.com/hrydgard/ppsspp/actions/runs/12236572156 would help, if not, then the performance issue is somewhere else

hrydgard commented:
I think you (or we, or anyone) also really should report this somehow as a bug to [VMA](https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator) ...  Though not exaclty sure how to describe it! Ideally we should have some smaller test case I suppose...

hrydgard commented:
Also, I should get around to upgrading VMA to latest, though I updated it not so long ago...

XStylus commented:
> @XStylus Thanks for testing! Perhaps https://github.com/hrydgard/ppsspp/actions/runs/12236572156 would help, if not, then the performance issue is somewhere else

Tested. Noticeable frame drops in Vulkan mode unfortunately still persist.

Thanks for at least sussing out the issue of Vulkan not working at all, at least.

hrydgard commented:
Hm, that's unfortunate, but I guess it is what it is. Vulkan performs great on the newer macs, at least.

XStylus commented:
> Hm, that's unfortunate, but I guess it is what it is. Vulkan performs great on the newer macs, at least.

Challenge accepted. 

I've just now tested on a brand new Mac Mini M4-Pro with 64GB memory running macOS 15.1.1.

Vulcan mode is smooth, no frame skip perceived. However, I did perceive image tearing. OpenGL mode was perfect. I have a video if you'd like.

That probably merits a second bug but I don't quite know how to describe it.

hrydgard commented:
@XStylus  have you tried the VSync setting? 

(I should probably replace that with an explicit list of the available presentation modes, btw...)

XStylus commented:
> @XStylus have you tried the VSync setting?

Well played, sir. I concede.

Kethen commented:
> > @XStylus have you tried the VSync setting?
> 
> Well played, sir. I concede. (Indeed, I had not, and that fixed the tearing on the Mac Mini M4.)
> 
> VSync unfortunately did not improve anything on my Macbook Pro, sadly. Quite the contrary, it slowed down performance tremendously. I would kindly suggest that it be a standing recommendation that macOS users with Radeon GPUs use OpenGL, despite the newfound Vulkan functionality for them from these patches.

@XStylus  Try updating 1.2.4 `PPSSPPSDL.app/Contents/Frameworks/libMoltenVK.dylib` with https://github.com/KhronosGroup/MoltenVK/releases/tag/v1.2.11-artifacts on the MacBook... tho on my rx480 both versions work fine

hrydgard commented:
Oh, good! Time to update the one we ship then.

Kethen commented:
linking https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator/issues/129 here for future reference
